components/script/task_queue.rs

```diff
@@ -0,0 +1,201 @@
+    pub fn recv(&self) -> Result<T, ()> {
+        self.internal.borrow_mut().pop_front().ok_or(())
+    }
+
+    // Same as recv.
+    pub fn try_recv(&self) -> Result<T, ()> {
+        self.recv()
+    }
+}
+
+// Defining the operations used to convert from a msg T to a QueuedTask.
+pub trait QueuedTaskConversion {
+    fn task_category(&self) -> Option<&ScriptThreadEventCategory>;
+    fn into_queued_task(self) -> Option<QueuedTask>;
+    fn from_queued_task(queued_task: QueuedTask) -> Self;
+    fn wake_up_msg() -> Self;
+    fn is_wake_up(&self) -> bool;
+}
+
+pub struct TaskQueue<T> {
+    // The original port on which the task-sources send tasks as messages.
+    port: Receiver<T>,
+    // A sender to ensure the port doesn't block on select while there are throttled tasks.
+    wake_up_sender: Sender<T>,
+    // A queue from which the event-loop can drain tasks.
+    msg_queue: MsgQueue<T>,
+    // A "business" counter, reset for each iteration of the event-loop
+    taken_task_counter: Cell<u64>,
+    // Tasks that will be throttled for as long as we are "busy".
+    throttled: DomRefCell<HashMap<TaskSourceName, VecDeque<QueuedTask>>>
+}
+
+#[allow(unsafe_code)]
+unsafe impl<T> JSTraceable for TaskQueue<T> {
+    #[allow(unsafe_code)]
+    unsafe fn trace(&self, _trc: *mut JSTracer) {
+        // Do nothing
+    }
+}
+
+impl<T: QueuedTaskConversion> TaskQueue<T> {
+    pub fn new(port: Receiver<T>, wake_up_sender: Sender<T>) -> TaskQueue<T> {
+        TaskQueue {
+            port,
+            wake_up_sender,
+            msg_queue: MsgQueue::new(),
+            taken_task_counter: Default::default(),
+            throttled: Default::default(),
+        }
+    }
+
+    // Process incoming tasks, immediately sending priority ones downstream,
+    // and categorizing potential throttles.
+    fn process_incoming_tasks(&self) {
+        let mut non_throttled: Vec<T> = self.port
+            .try_iter()
+            .filter(|msg| !msg.is_wake_up())
+            .collect();
+
+        let to_be_throttled: Vec<T> = non_throttled.drain_filter(|msg|{
+            let category = match msg.task_category() {
+                Some(category) => category,
+                None => return false,
+            };
+            match category {
+                ScriptThreadEventCategory::PerformanceTimelineTask => return true,
+                _ => {
+                    // A task that will not be throttled, start counting "business"
+                    self.taken_task_counter.set(self.taken_task_counter.get() + 1);
+                    return false
+                },
+            }
+        }).collect();
+
+        for msg in non_throttled {
+            // Immediately send non-throttled tasks for processing.
+            let _ = self.msg_queue.send(msg);
+        }
+
+        for msg in to_be_throttled {
+            // Categorize tasks per task queue.
+            let (worker, category, boxed, pipeline_id) = msg.into_queued_task().unwrap();
```

Comment:
I changed it to a match with `unreachable`. 

I think there is a contract that a message whose category is non-priority, hence was included in `to_be_throttled` should implement `into_queued_task` that returns `Some`, otherwise it would point to a logical error either in the `filter` above, or in the implementation of `QueuedTaskConversion` for that message type. 

We still need an option, because not all variants of the message represent something that could be converted into a queued task(this was basically the main difficulty of getting to a generic task queue, since the variant containing a `CommonScriptMsg::Task` is wrapped into different enum with different structures).